### ISSUE TITLE
findFailure fails to find the correct failure in a more complicated case

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2803]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)| @Peaker reported|
|-|-|
|Trac ID|trac#2803|
|Type|defect|
|Created|2007-08-25 07:08:08Z|
|Branch|https://github.com/twisted/twisted/tree/findfailure-2803|

The problem is that findFailure uses the fact that the caller of the frame in which except: was used is "throwExceptionIntoGenerator" but that doesn't necessarily mean that we are actually catching *that*. We may be catching something that was re-raised from a catcher of that exception.

I think a solution would be to verify that it was indeed a yield that raised it, and not only that throwExceptionIntoGenerator is the caller frame.

I attached a patch with an extra test in test_failure's generator_tests to demonstrate the problem.

Attachments:

* [test_failure.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/422/42271387fd9aeff9e75d653d2f08c7da3e3d8ec6/59a7dbbc008889b47b76c6c1a02e07e6f2d12b5a.patch) (2585 bytes) - added by Peaker on 2007-08-25 07:08:36Z - A test to demonstrate a problem with findFailure

<details><summary>Searchable metadata</summary>

```
trac-id__2803 2803
type__defect defect
reporter__Peaker Peaker
priority__highest highest
milestone__ 
branch__branches_findfailure_2803 branches/findfailure-2803
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1188025688000000 1188025688000000
changetime__1195842436000000 1195842436000000
version__None None
owner__Peaker Peaker
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_1"></a>|@Peaker commented|
|-|-|

I created a branch findfailure-2803 for this ticket, so ignore the patch file.

**Comment by twisted-trac** [comment]
|<img alt="peaker's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_2"></a>|peaker commented|
|-|-|

(In [21923]) Naming conventions, spacing conventions, and a docstring. refs [#2803](https://github.com/twisted/twisted/issues/2803)

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_3"></a>|@Peaker removed owner|
|-|-|

Hairy situation requires hairy fixes!

I think its ready for review...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve set owner to @Peaker|
|-|-|

 * _yieldOpcode is unnecessary, ```opcode.opmap["YIELD_VALUE"]``` exists in python 2.3. I suggest making the value a private class variable of Failure (but with documentation).
 * several whitespaces missing around ```=``` operator in tests
 * instead of ```self.assertEqual(Failure().type, IndexError)```, I prefer ```self.assertIsInstance(Failure().value, IndexError)```
 * in the docstrings of ```test_ambiguousFailureInGenerator``` and ```test_ambiguousFailureFromGenerator```, please use ```L{Failure._findFailure```} instead of findFailure
 * trailing whitespace in failure.py

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_5"></a>|@Peaker set owner to @therve|
|-|-|

Replying to [therve](#note_4):

Thanks for the review.

>  * _yieldOpcode is unnecessary, ```opcode.opmap["YIELD_VALUE"]``` exists in python 2.3. I suggest making the value a private class variable of Failure (but with documentation).

Ah, didn't know that. Done.

>  * several whitespaces missing around ```=``` operator in tests
Fixed.

>  * instead of ```self.assertEqual(Failure().type, IndexError)```, I prefer ```self.assertIsInstance(Failure().value, IndexError)```

Done.

>  * in the docstrings of ```test_ambiguousFailureInGenerator``` and ```test_ambiguousFailureFromGenerator```, please use ```L{Failure._findFailure```} instead of findFailure

Done.

>  * trailing whitespace in failure.py

That wasn't mine, but fixed anyhow.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @Peaker|
|-|-|

OK please merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve set status to `closed`|
|-|-|

(In [21945]) Merge findfailure-2803

Author: peaker
Reviewer: therve
Fixes [#2803](https://github.com/twisted/twisted/issues/2803)

Fix a corner case in Failure._findFailure where throwExceptionIntoGenerator is
in the caller frame but not the real exception.
